src/mlpack/methods/ann/layer/nearest_interpolation_impl.hpp

```diff
@@ -0,0 +1,161 @@
+                     const size_t inColSize,
+                     const size_t outRowSize,
+                     const size_t outColSize,
+                     const size_t depth) :
+    inRowSize(inRowSize),
+    inColSize(inColSize),
+    outRowSize(outRowSize),
+    outColSize(outColSize),
+    depth(depth),
+    batchSize(0)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void NearestInterpolation<InputDataType, OutputDataType>::Forward(
+  const arma::Mat<eT>& input, arma::Mat<eT>& output)
+{
+  batchSize = input.n_cols;
+  if (output.is_empty())
+    output.set_size(outRowSize * outColSize * depth, batchSize);
+  else
+  {
+    assert(output.n_rows == outRowSize * outColSize * depth);
+    assert(output.n_cols == batchSize);
+  }
+
+  assert(inRowSize >= 2);
+  assert(inColSize >= 2);
+
+  arma::cube inputAsCube(const_cast<arma::Mat<eT>&>(input).memptr(),
+      inRowSize, inColSize, depth * batchSize, false, false);
+  arma::cube outputAsCube(output.memptr(), outRowSize, outColSize,
+      depth * batchSize, false, true);
+
+  double scaleRow = (double) inRowSize / (double) outRowSize;
+  double scaleCol = (double) inColSize / (double) outColSize;
+
+  for (size_t i = 0; i < outRowSize; ++i)
+  {
+    const size_t rOrigin = std::floor(i * scaleRow);
+
+    for (size_t j = 0; j < outColSize; ++j)
+    {
+      const size_t cOrigin = std::floor(j * scaleCol);
+
+      for (size_t k = 0; k < depth * batchSize; ++k)
+      {
+        outputAsCube(i, j, k) = inputAsCube.slice(k)(
+            rOrigin, cOrigin);
+      }
+    }
+  }
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void NearestInterpolation<InputDataType, OutputDataType>::Backward(
+  const arma::Mat<eT>& /*input*/,
+  const arma::Mat<eT>& gradient,
+  arma::Mat<eT>& output)
+{
+  if (output.is_empty())
+  {
+    output.set_size(inRowSize * inColSize * depth, batchSize);
+  }
+  else
+  {
+    assert(output.n_rows == inRowSize * inColSize * depth);
+    assert(output.n_cols == batchSize);
+  }
+
+  assert(outRowSize >= 2);
+  assert(outColSize >= 2);
+
+  arma::cube outputAsCube(output.memptr(), inRowSize, inColSize,
+      depth * batchSize, false, true);
+  arma::cube gradientAsCube(((arma::Mat<eT>&) gradient).memptr(), outRowSize,
+      outColSize, depth * batchSize, false, false);
+
+  double scaleRow = (double)(inRowSize) / outRowSize;
+  double scaleCol = (double)(inColSize) / outColSize;
+
+  if (gradient.n_elem == output.n_elem)
+  {
+    outputAsCube = gradientAsCube;
+  }
+  else
+  {
+    for (size_t i = 0; i < outRowSize; ++i)
+    {
+      const size_t rOrigin = std::floor(i * scaleRow);
+
+      for (size_t j = 0; j < outColSize; ++j)
+      {
+        const size_t cOrigin = std::floor(j * scaleCol);
+
+        for (size_t k = 0; k < depth * batchSize; ++k)
+        {
+          outputAsCube(rOrigin, cOrigin, k) +=
```

Comment:
This is undefined behavior as well, there is no guarantee that output is initialized with zero.

Reply:
If the output is empty then I used `set_size` to give its size and by default the elements are initialised to 0.
Still I will `.zeros` just to be safe.

Reply:
That depends on the system:

> in Armadillo 10.4 and earlier versions, the elements are not initialised during construction unless fill_form is specified
> (ie. without specifying fill_form, the elements may contain garbage values, including NaN)
> in Armadillo 10.5 and later versions, by default the elements are initialised to zero during construction

Since the CI uses 8.400.0, it's not guaranteed that it's zero.